bin/apply

```diff
@@ -65,23 +58,43 @@ if (@thingies) { # Ignore $argc, found our own.
 }
 
 # Now, apply the command till we run out.
+my $err = EX_SUCCESS;
 while (@ARGV && @ARGV >= $argc) {
     if (@thingies) {
        (my $new_command = $command) =~ s/${magic}(\d+)/$ARGV [$1 - 1]/ge;
-        system $new_command;   # Reinterpreted by the shell!
+        my $rc = system $new_command;   # Reinterpreted by the shell!
+        checkerr($rc);
         splice @ARGV, 0, $argc;
     }
     else {
         if ($argc) {
-            system $command, splice @ARGV, 0, $argc;
+            my $rc = system $command, splice @ARGV, 0, $argc;
+            checkerr($rc);
         }
         else {
-            system $command;
+            my $rc = system $command;
+            checkerr($rc);
             shift;
         }
     }
 }
-exit EX_SUCCESS;
+exit $err;
+
+sub checkerr {
+    my $status = shift;
+    if ($status != 0) {
```

Comment:
This is a little weird, because not every command returns 0 for success. Some commands have more than one success value. 

If this is the way actual `apply` works though, we can just copy that even if it misses edge cases.

Reply:
In the OpenBSD code, running the command is handled in the function mysystem() [1].
If fork() fails, mysystem() returns -1.
If fork() succeeds and execl() fails in the child process, the child will exit with status 1 (fetched via waitpid).
If execl() succeeds in the child process, but then waitpid() fails, pid is updated and mysystem() returns -1.
Otherwise, mysystem() returns the value from waitpid().

Return value of 0 from mysystem() is treated as "success".
Since it is called in a loop, if any instance of mysystem() returns non-zero, apply will do exit(1).
It doesn't matter if the executed command used exit code 1 or 100, apply will exit with 1.
The FreeBSD, NetBSD and OpenBSD manuals don't appear to mention this.

In the perl code, if system() returned -1 then likely either fork() or waitpid() failed.
Printing an error here is close to the C version because the command could not be started for some reason (e.g. not found).
An error is not printed if the command had exit code 100, because this information is passed on *lossily* by the eventual exit code of 1.

1. https://cvsweb.openbsd.org/cgi-bin/cvsweb/src/usr.bin/apply/apply.c?annotate=1.29

Reply:
Thanks for the code pointers. I think that code is being less thoughtful than it could be, but that doesn't matter because we are copying what it does and not what we think it should do.